x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/assignment/planning/ZoneAwareAssignmentPlanner.java

```diff
@@ -0,0 +1,200 @@
+        if (nodesByZone.size() == 1) {
+            return new AssignmentPlanner(nodesByZone.values().iterator().next(), models).computePlan(true);
+        }
+
+        // First we try to compute a plan without forcing assigning previously assigned models as this may
+        // produce better plans. If that plan has failed to assign previously assigned models we then try
+        // again this time prioritizing assigning such models.
+        AssignmentPlan plan = computePlan(false);
+        if (plan.arePreviouslyAssignedModelsAssigned() == false) {
+            plan = computePlan(true);
+        }
+        return plan;
+    }
+
+    private AssignmentPlan computePlan(boolean tryAssigningPreviouslyAssignedModels) {
+        logger.debug(
+            () -> format(
+                "computing plan %s trying to assign previously assigned models",
+                tryAssigningPreviouslyAssignedModels ? "" : "without"
+            )
+        );
+        // The idea here is that we solve per zone trying to distribute allocations evenly.
+        // After computing a plan for each zone it is possible that there are still unsatisfied allocations
+        // that can be allocated, so we solve a final time across all zones preserving the allocations we
+        // allocated on the first per zone assignment plans.
+
+        int remainingZones = nodesByZone.size();
+        Map<String, Integer> modelIdToRemainingAllocations = models.stream().collect(Collectors.toMap(Model::id, Model::allocations));
+        List<AssignmentPlan> plans = new ArrayList<>();
+        for (var zoneToNodes : nodesByZone.entrySet()) {
+            logger.debug(() -> format("computing plan for availability zone %s", zoneToNodes.getKey()));
+            AssignmentPlan plan = computeZonePlan(
+                zoneToNodes.getValue(),
+                modelIdToRemainingAllocations,
+                remainingZones,
+                tryAssigningPreviouslyAssignedModels
+            );
+            plan.models()
+                .forEach(
+                    m -> modelIdToRemainingAllocations.computeIfPresent(
+                        m.id(),
+                        (modelId, remainingAllocations) -> remainingAllocations - plan.totalAllocations(m)
+                    )
+                );
+            plans.add(plan);
+            remainingZones--;
+        }
+        AssignmentPlan plan = computePlanAcrossAllNodes(plans);
+        logger.debug(() -> "Zone aware plan =\n" + plan.prettyPrint());
+        return plan;
+    }
+
+    private AssignmentPlan computeZonePlan(
+        List<Node> nodes,
+        Map<String, Integer> modelIdToRemainingAllocations,
+        int remainingZones,
+        boolean tryAssigningPreviouslyAssignedModels
+    ) {
+        Map<String, Integer> modelIdToTargetAllocations = modelIdToRemainingAllocations.entrySet()
+            .stream()
+            .collect(Collectors.toMap(e -> e.getKey(), e -> (e.getValue() - 1) / remainingZones + 1));
```

Comment:
There's an interesting effect here.

Suppose we have 3 AZs, some models that have been given 3 allocations and some jobs that have been given 1 allocation.

We'll try to give all the jobs 1 allocation in the first AZ.  This likely won't fit, but it could be that we end up choosing the models that are going to have 1 allocation in total.  In the worst case we could end up with all the models with 1 allocation in the first AZ, and the models with 3 allocations having to have 1 and 2 allocations in the other two AZs.  This would be highly unintuitive to a user who would expect that the models they've asked for 3 allocations of to go one per AZ, and the ones they've asked for 1 allocation of to be fitted in around that.

One possibility might be to exclude models with only one allocation requested from the per-zone plans, so they get fitted in at the very end by the overall plan.  If you think this is too hard for the time available for 8.5 then I am happy to leave this as-is for 8.5.  I think what's here at the moment is better than what we have now, so the PR in its current form is a step forwards.  But we might need to revisit this again in the future.

Reply:
Right, yes, that is possible. However, if we excluded 1-allocation models from the per-zone planning, then it is possible that they do not get any allocations if the multi-allocation models took all resources. It would be like tweaking the objective function to prefer multi-allocation models. Not sure if that is better.